### PR TITLE
FileSystemLocal adapter plugin

### PR DESCRIPTION
**Pull request by kasvith**
Pull Request for Issue #159  .

### Summary of Changes
- Moved local adapter from **com_media** to a plugin  **FileSystemLocal** 

### Testing Instructions
- Discover and Enable plugin if it is not enabled yet
- Open Media Manager

### Expected result
- Working Media Manager with new Local FileSystem plugin

### Actual result
- Working local filesystem with new plugin

### CODE REVIEW

File: administrator/components/com_content/dispatcher.php [hunk 1]
```diff
@@ -28,7 +28,8 @@ class ContentDispatcher extends Dispatcher
 	 *
 	 * @since   __DEPLOY_VERSION__
 	 */
-	public function __construct($namespace, \Joomla\CMS\Application\CmsApplication $app, \JInput $input = null, \Joomla\CMS\Mvc\Factory\MvcFactoryInterface $factory = null)
+	public function __construct($namespace, \Joomla\CMS\Application\CmsApplication $app, \JInput $input = null,
```
Comment (laoneo) [inline administrator/components/com_content/dispatcher.php]: Please revert that change.

File: plugins/filesystem/local/adapter/adapter.php [hunk 2]
```diff
@@ -81,6 +84,98 @@ public function getFile($path = '/')
 	}
 
 	/**
+	 * Returns the folder or file information for the given path. The returned object
```
Comment (laoneo) [inline plugins/filesystem/local/adapter/adapter.php]: Why did you change the whole block here?

File: plugins/filesystem/local/adapter/adapter.php [hunk 3]
```diff
@@ -12,6 +12,9 @@
 JLoader::import('joomla.filesystem.file');
 JLoader::import('joomla.filesystem.folder');
 
+require_once 'filenotfoundexception.php';
```
Comment (laoneo) [inline plugins/filesystem/local/adapter/adapter.php]: These files need to stay in com_media as they are the same for all future cloud plugins.

**Review by laoneo — CHANGES_REQUESTED** [review]
How does the api model get then the adapter at all?

**Comment by laoneo** [comment]
A good start, but needs some changes.

File: plugins/filesystem/local/adapter/adapter.php (hunk 2 above)
Comment (kasvith) [reply plugins/filesystem/local/adapter/adapter.php]: what you meant ?

File: plugins/filesystem/local/adapter/adapter.php (hunk 3 above)
Comment (kasvith) [inline plugins/filesystem/local/adapter/adapter.php]: is there a joomla way to import them ?

File: administrator/components/com_content/dispatcher.php (hunk 1 above)
Comment (kasvith) [reply administrator/components/com_content/dispatcher.php]: Done, did due to build fail

File: plugins/filesystem/local/adapter/adapter.php (hunk 3 above)
Comment (kasvith) [inline plugins/filesystem/local/adapter/adapter.php]: or can i use require_once for that, if it does not violating Joomla! rules

Comment (kasvith) [inline plugins/filesystem/local/adapter/adapter.php]: will move it to com_media and import to this file

Comment (kasvith) [reply plugins/filesystem/local/adapter/adapter.php]: Anyway if these are common to cloud filesystems
Why we don't use this plugin as the one who really deals with files, like PlgFileSystemGoogleDrive does not need to know how a file handled in Local File System, it is encapsulated on this plugin, it can provide the functionality without a change.

Any comments about that approach ?
I moved those keeping that design in my mind

File: plugins/filesystem/local/adapter/adapter.php (hunk 2 above)
Comment (kasvith) [reply plugins/filesystem/local/adapter/adapter.php]: There were some code syntax issues in this file, in incorrect format. I fixed them to Joomla! standard

there was like

if(){

}

statements and changed them to

if()
{

}

**Comment by kasvith** [comment]
@laoneo oops sorry after a pull request my code has been changed, thanks for mentioning it above 👍

File: plugins/filesystem/local/adapter/adapter.php (hunk 3 above)
Comment (laoneo) [reply plugins/filesystem/local/adapter/adapter.php]: But the api model needs to know what it gets when it asks for an adapter, so the interface needs to stay there. com_media should not have any dependency to a plugin but the plugins to com_media.

File: plugins/filesystem/local/adapter/adapter.php (hunk 2 above)
Comment (laoneo) [reply plugins/filesystem/local/adapter/adapter.php]: If you look on the code removed and added, it's the same.

File: plugins/filesystem/local/adapter/adapter.php (hunk 3 above)
Comment (kasvith) [reply plugins/filesystem/local/adapter/adapter.php]: will fix that ASAP

File: plugins/filesystem/local/adapter/adapter.php (hunk 2 above)
Comment (kasvith) [reply plugins/filesystem/local/adapter/adapter.php]: oh i see, i think i've used space before, then changed back to tabs
Can you mark it as resolved :)

**Comment by kasvith** [comment]
@laoneo can you check now ?

**Comment by laoneo** [comment]
Please move the interfaces back to their original place. Thanks.

**Comment by kasvith** [comment]
Moved interfaces to *com_media/libraries/media/file/adapter*

File: plugins/filesystem/local/adapter/adapter.php (hunk 2 above)
Comment (laoneo) [reply plugins/filesystem/local/adapter/adapter.php]: Did you modify something in that code? There is for me no reason why it should be changed. For me it looks identical.

**Review by laoneo — CHANGES_REQUESTED** [review]

File: plugins/filesystem/local/adapter/adapter.php [hunk 4]
```diff
@@ -12,6 +12,13 @@
 JLoader::import('joomla.filesystem.file');
 JLoader::import('joomla.filesystem.folder');
 
+// Path for Interfaces in com_media
+define('MEDIA_INTERFACE_PATH', JPATH_ADMINISTRATOR . '/components/com_media/libraries/media/file/adapter/');
+
+include_once MEDIA_INTERFACE_PATH . 'interface.php';
```
Comment (laoneo) [inline plugins/filesystem/local/adapter/adapter.php]: They should be loaded trough JLoader like here https://github.com/joomla-projects/media-manager-improvement/blob/milestone1/plugins/media-action/crop/crop.php#L12

**Comment by kasvith** [comment]
@laoneo it is changing **spaces** to **tabs** caused that problem

**Comment by laoneo** [comment]
The original file doesn't have spaces.

**Comment by kasvith** [comment]
yes, but when i imported file to PHPStrom it asked me to switch to tabs, so i switched at first time and commited

File: plugins/filesystem/local/adapter/adapter.php (hunk 2 above)
Comment (kasvith) [reply plugins/filesystem/local/adapter/adapter.php]: I first commited to my repo with spaces, may that be the cause ?

**Comment by kasvith** [comment]
Used **JLoader**

File: plugins/filesystem/local/adapter/adapter.php (hunk 2 above)
Comment (laoneo) [reply plugins/filesystem/local/adapter/adapter.php]: Just copy it from the original file and then the changes should disappear.

File: plugins/filesystem/local/local.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ * @package     Joomla.Plugin
+ * @subpackage  FileSystem.Local
+ *
+ * @copyright   Copyright (C) 2005 - 2017 Open Source Matters, Inc. All rights reserved.
+ * @license     GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
+defined('_JEXEC') or die;
+
+require_once 'adapter/adapter.php';
```
Comment (laoneo) [inline plugins/filesystem/local/local.php]: Here also JLoader

File: administrator/components/com_media/models/api.php [hunk 6]
```diff
@@ -9,6 +9,7 @@
 
 defined('_JEXEC') or die;
 
+
```
Comment (laoneo) [inline administrator/components/com_media/models/api.php]: This one here is not needed

File: plugins/filesystem/local/adapter/adapter.php [hunk 7]
```diff
@@ -11,7 +11,8 @@
 
 JLoader::import('joomla.filesystem.file');
 JLoader::import('joomla.filesystem.folder');
-
+JLoader::import('components.com_media.libraries.media.file.adapter.interface', JPATH_ADMINISTRATOR);
+JLoader::import('components.com_media.libraries.media.file.adapter.filenotfoundexception', JPATH_ADMINISTRATOR);
 /**
```
Comment (laoneo) [inline plugins/filesystem/local/adapter/adapter.php]: Here is a space needed.

**Review by laoneo — CHANGES_REQUESTED** [review]

File: plugins/filesystem/local/local.php (hunk 5 above)
Comment (kasvith) [reply plugins/filesystem/local/local.php]: with same method ?

Comment (laoneo) [reply plugins/filesystem/local/local.php]: Yes

Comment (kasvith) [reply plugins/filesystem/local/local.php]: done

File: administrator/components/com_media/models/api.php (hunk 6 above)
Comment (kasvith) [reply administrator/components/com_media/models/api.php]: using CMS over cms ?

**Comment by laoneo** [comment]
Thank you @kasvith, good work!
